### PR TITLE
[PM-8217] New device two factor notice

### PR DESCRIPTION
**Pull request by andrebispo5**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-8217

## 📔 Objective

Adds navigation to the new device two factor notice pages.
Includes logic to have a display status linked to the current user.
Adds and updates tests.
Remind me later button is hidden by permanent feature flag. 
## Screenshot
Permanent feature flag on:
image

Permanent feature flag off:
image

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt [hunk 1]
```diff
@@ -1333,6 +1337,80 @@ class AuthRepositoryImpl(
...
+        if (!isCloudUser || !isProfileAtLeastWeekOld) {
+            return false
+        }
+
+        if (hasSSOPolicy || hasTwoFactorEnabled) {
+            return false
+        }
+
+        return true
+    }
```
Comment (fedemkr) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1419]: 🎨  Up to you but IMO you could short-circuit it to return in several parts this so you don't need to waste resources fetching other stuff. Like there's no need to fetch the policies if the user is not cloud.
```suggestion
    private fun newDeviceNoticePreConditionsValid(): Boolean {
        if (environmentRepository.environment.type == Environment.Type.SELF_HOSTED) {
            return false
        }

        val userProfile = authDiskSource.userState?.activeAccount?.profile
        val isProfileAtLeastWeekOld = userProfile
            ?.let {
                it.creationDate
                    ?.plusWeeks(1)
                    ?.isBefore(
                        ZonedDateTime.now(),
                    )
            }
            ?: false
        if (!isProfileAtLeastWeekOld) {
            return false
        }

        val hasTwoFactorEnabled = userProfile
            ?.isTwoFactorEnabled
            ?: false
        if (hasTwoFactorEnabled) {
            return false
        }

        val hasSSOPolicy =
            policyManager.getActivePolicies(type = PolicyTypeJson.REQUIRE_SSO)
                .any { p -> p.isEnabled }

        return !hasSSOPolicy
    }
```

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/model/NewDeviceNoticeDisplayStatus.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ */
+@Suppress("MagicNumber")
+@Serializable
+data class NewDeviceNoticeState(
+    @SerialName("displayStatus")
+    val displayStatus: NewDeviceNoticeDisplayStatus,
+
+    @SerialName("delayDate")
+    @Contextual
+    val delayDate: ZonedDateTime?,
```
Comment (fedemkr) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/model/NewDeviceNoticeDisplayStatus.kt:50]: 🤔 Should this be called `lastSeenDate`? Maybe it's just me but I would connect "delay" to something that I need to add to a date that I need to wait for and from what I can see in the code this date is referring to the moment the user saw the notice.

File: app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/newdevicenotice/NewDeviceNoticeEmailAccessViewModel.kt [hunk 3]
```diff
@@ -38,8 +47,24 @@ class NewDeviceNoticeEmailAccessViewModel @Inject constructor(
...
+                if (featureFlagManager.getFeatureFlag(FlagKey.NewDevicePermanentDismiss)) {
+                    CAN_ACCESS_EMAIL_PERMANENT
+                } else {
+                    CAN_ACCESS_EMAIL
+                }
+
+            authRepository.setNewDeviceNoticeState(
+                NewDeviceNoticeState(
+                    displayStatus = displayStatus,
+                    delayDate = ZonedDateTime.now(),
```
Comment (fedemkr) [inline app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/newdevicenotice/NewDeviceNoticeEmailAccessViewModel.kt:61]: 🤔 Maybe I'm misunderstanding the views/flows, but shouldn't `delayDate` be `nil` here given the status is not `HAS_SEEN`?

File: app/src/test/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModelTest.kt [hunk 4]
```diff
@@ -1333,6 +1334,45 @@ class RootNavViewModelTest : BaseViewModelTest() {
         )
     }
 
+    @Suppress("MaxLineLength")
+    @Test
+    fun `when the active user has an unlocked vault and they need to be shown the new device notice the nav state should be NewDeviceTwoFactorNotice`() {
+        every { authRepository.checkUserNeedsNewDeviceTwoFactorNotice() } returns true
+        mutableUserStateFlow.tryEmit(
```
Comment (fedemkr) [inline app/src/test/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModelTest.kt:1341]: 🤔 I think another test should be added where `authRepository.checkUserNeedsNewDeviceTwoFactorNotice` returns `false` to make sure the navigation goes to the proper place as well.

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/model/NewDeviceNoticeDisplayStatus.kt [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+) {
+    /**
+     * Whether the [delayDate] is at least 7 days old.
+     */
+    val delayDateIsSevenDaysOld = delayDate?.isBefore(ZonedDateTime.now().minusDays(7)) ?: false
```
[the remark this answers is not in the document]
Comment (andrebispo5) [reply app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/model/NewDeviceNoticeDisplayStatus.kt:55]: No constant file for single place usage.

File: app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/newdevicenotice/NewDeviceNoticeEmailAccessViewModel.kt (hunk 3 above)
Comment (andrebispo5) [reply app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/newdevicenotice/NewDeviceNoticeEmailAccessViewModel.kt:61]: You are right the date here is irrelevant

File: app/src/test/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModelTest.kt (hunk 4 above)
Comment (andrebispo5) [reply app/src/test/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModelTest.kt:1341]: yup makes sense

**Review by fedemkr — APPROVED** [review]

**Review by differsthecat — APPROVED** [review]

**Pull request merged**